---
title: "Supplementary Figure 11"
author: "Damian Wollny"
output:
  html_document:
    theme: paper
    df_print: kable
date: '`r format(Sys.Date(), "%B %d, %Y")`'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = FALSE, message = FALSE)
```

```{r}
# Setup
library(tidyverse)
library(mgcv)
library(ggseqlogo)
library(Biostrings)
library(seqinr)

theme_set(theme_classic(base_size=24))
```

<style type="text/css">

body, td {
   font-size: 18px;
}
</style>

## Figure S11a

```{r}
# Get MEF data
ool41 <- readRDS(file = "../_input/mef_data.RDS") %>% as.tbl()

# Clean up MEF data
ool41_droplets <- ool41 %>% 
  as.tbl() %>% 
  filter(grepl(pattern = "ool41*", .$cell_id)) %>% 
  # kickout + ctrl (1000 droplets) & - ctrl (0 droplets)
  filter(cell_id != "ool41_p1_a1",
         cell_id != "ool41_p2_a1",
         cell_id != "ool41_p1_h12",
         cell_id != "ool41_p2_h12")

input_tpm <- ool41 %>%
  filter(cell_id == "mef_5ng") %>% 
  filter(tpm != 0) %>% 
  group_by(target_id) %>% 
  summarise(input_tpm = log2_tpm)

mef_bens_plot <- ool41_droplets %>% 
  filter(tpm != 0) %>% 
  group_by(target_id) %>% 
  summarise(ave_tpm = mean(log2_tpm), count = n()) %>% 
  inner_join(.,input_tpm, by="target_id") %>% distinct() %>% 
  arrange(desc(count)) %>% 
  mutate(percent = count*100/max(count)) 

# get residuals from model describing data in bens_plot
mef_bens_plot_model <- gam(percent ~ s(input_tpm, bs = "cs"), data = mef_bens_plot)
mef_bens_plot_5ng <- mef_bens_plot %>% mutate(residuals = mef_bens_plot_model$residuals)

bens_plot_5ng <- mef_bens_plot_5ng %>% 
  ggplot(aes(x = input_tpm, y = percent)) +
  geom_point(size = 3, pch = 21) +
  geom_point(data = mef_bens_plot_5ng %>% filter(residuals > 30), color = "indianred4", size = 4, alpha = 0.4) +
  stat_smooth(col = "red") +
  labs(x = "Input mRNA amount\nlog2(TPM)", y = "Detected in x\ncoacervates [%]") +
  ylim(c(0,100))
plot(bens_plot_5ng)
```

## Figure S11b

Prior to this script: run 'meme_parser_prob_mat.py' to get input files in the right format for the ggseqlogo package

```{r, fig.height=12}
# great vignette for ggseqlogo found on: https://omarwagih.github.io/ggseqlogo/
# function to format parsed meme output
get_in_shape <- function(file){
  the_data <- read.table(file = file, sep = " ", skip = 1)
  mat <- rbind(the_data$V2,the_data$V4, the_data$V6, the_data$V8)
  rownames(mat) <- c("A", "C", "G", "T")
  return(mat)
}

# load all motif_*.txt files and apply the get_in_shape function
mat_list <- lapply(Sys.glob("./_meme/motif_*.txt"), get_in_shape)

# correct mat_list order and rename
mat_list <- mat_list[c(1,3,4,5,6,7,8,9,10,2)]
names(mat_list) <- paste("Motif:", c(1:10))

# Generate sequence logo
mef_motifs <- ggseqlogo(mat_list[1:8], method = "bits", font = "helvetica_regular", ncol = 1) + theme(strip.text.x = element_text(size=22)) 
mef_motifs
```

## Figure S11c

Run 'meme_parser_get_ind_motifs.py' on MEME output to get the input file for the following plot

```{r}
# function to get the reverse complement of given seq
get_reverse_compl <- function(seq){
  reverse(chartr("ATGC","TACG",seq))  
}

# Read Sine sequences obtained from RepBase
sines <- read.fasta("./_sines/mousub_gatc.fa")

# Parse fasta into df
sine_df <- data.frame(sine = character(),
                      sine_seq = character(),
                      stringsAsFactors=FALSE)

for (i in 1:length(sines)){
  sine_df[i,1] <- names(sines[i])
  sine_df[i,2] <- toupper(c2s(sines[[i]]))
}

# Read individual motifs from meme analysis - ENRICHED
ind_motifs_enr <- read.csv(file = "./_sines/meme_ind_motifs.csv")
ind_motifs_enr_top <- ind_motifs_enr %>% 
  group_by(motif_number) %>% 
  filter(row_number() <= 5) %>% 
  modify_if(is.factor, as.character) %>% 
  as.data.frame() %>% 
  mutate(rc_motif_seqs = get_reverse_compl(motif_seqs))

# Define output df
output_df <- data.frame(sine = character(),
                        motif = character(),
                        align_score = numeric(),
                        motif_source = character(),
                        stringsAsFactors = FALSE)

# Setup Substitution matrix for alignment
sigma <- nucleotideSubstitutionMatrix(match = 2, mismatch = -1, baseOnly = TRUE)

x = 0
for (j in 1:nrow(sine_df)){
  for (n in 1:nrow(ind_motifs_enr_top)) {
    x = x + 1
    output_df[x,1] <- sine_df$sine[j] 
    output_df[x,2] <- ind_motifs_enr_top[n,1]
    output_df[x,3] <- pairwiseAlignment(sine_df[j,2], as.character(ind_motifs_enr_top[n,3]), substitutionMatrix = sigma, gapOpening = -10,
                      gapExtension = 0, scoreOnly = TRUE)
    output_df[x,4] <- "enriched"
  }
}

# plot
output_df %>% 
  filter(motif_source == "enriched",
         sine == "B1_Mm" | sine == "B1_Mus1") %>% 
  mutate(motif = factor(motif, levels = rev(c("MEME-1", "MEME-2", "MEME-3", "MEME-4", "MEME-5",
                                          "MEME-6", "MEME-7", "MEME-8", "MEME-9", "MEME-10")))) %>% 
  group_by(sine,motif) %>% 
  summarise(ave_score = mean(align_score)) %>% 
  ggplot(aes(x = sine, y = motif, fill = ave_score)) +
  geom_tile(color = "white") +
  scale_fill_distiller(palette="Blues", direction=1, values = rev(c(1.0,0.8,0.75,0.7,0.6,0))) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) 
```

  